File: priv/deontic/deontic.exs
# SPEC_deontic_core — keel-language embodiment (Elixir), exhaustively verified.
# Mirrors deontic.py + precedence.py: derivation, De Morgan chain, fold/resolve,
# weighted compile, precedence ladder. Pattern-matched, immutable, zero deps.
# Run: elixir deontic.exs  (exit 0 iff all laws hold)
#
# Vocabulary note: stated in neutral deontic-logic terms because it is applied
# to isomorphisms internal to this system. Structural coincidence with any
# external normative tradition is a remark, not a claim.

defmodule Deontic do
  # -- Derivation: three axes -> exactly five inhabitants -----------------
  @axes [{:demand, :act, true}, {:demand, :act, false},
         {:demand, :refrain, true}, {:demand, :refrain, false}, {:option}]
  def name({:demand, :act, true}),       do: :required
  def name({:demand, :act, false}),      do: :encouraged
  def name({:demand, :refrain, true}),   do: :forbidden
  def name({:demand, :refrain, false}),  do: :discouraged
  def name({:option}),                   do: :neutral
  def axes, do: @axes
  def h, do: Enum.map(@axes, &name/1)

  # -- Valence chain and involution dual ----------------------------------
  def v(:forbidden), do: 0.0
  def v(:discouraged), do: 0.25
  def v(:neutral), do: 0.5
  def v(:encouraged), do: 0.75
  def v(:required), do: 1.0
  def dual(:required), do: :forbidden
  def dual(:encouraged), do: :discouraged
  def dual(:neutral), do: :neutral
  def dual(:discouraged), do: :encouraged
  def dual(:forbidden), do: :required

  # -- Aggregation monoid M and resolution --------------------------------
  def e(:required), do: {2, 0}
  def e(:encouraged), do: {1, 0}
  def e(:neutral), do: {0, 0}
  def e(:discouraged), do: {0, 1}
  def e(:forbidden), do: {0, 2}
  def fold({c1, o1}, {c2, o2}), do: {max(c1, c2), max(o1, o2)}
  def resolve({2, 2}), do: :conflict
  def resolve({2, _}), do: :required
  def resolve({_, 2}), do: :forbidden
  def resolve({1, 1}), do: :neutral
  def resolve({1, 0}), do: :encouraged
  def resolve({0, 1}), do: :discouraged
  def resolve({0, 0}), do: :neutral
  def combine(hs), do: hs |> Enum.map(&e/1) |> Enum.reduce({0, 0}, &fold/2) |> resolve()

  # -- Weighted claims: lex-max on {strength, weight} (gap 3) -------------
  def lmax(x, y), do: if(x >= y, do: x, else: y)
  def wclaim_fold(x, y), do: {lmax(elem(x, 0), elem(y, 0)), lmax(elem(x, 1), elem(y, 1))}
  def resolve_w({{2, _}, {2, _}}), do: :conflict
  def resolve_w({{2, _}, _}), do: :required
  def resolve_w({_, {2, _}}), do: :forbidden
  def resolve_w({{1, cw}, {1, ow}}) when cw > ow, do: :encouraged
  def resolve_w({{1, cw}, {1, ow}}) when ow > cw, do: :discouraged
  def resolve_w({{1, _}, {1, _}}), do: :neutral_gated
  def resolve_w({{1, _}, _}), do: :encouraged
  def resolve_w({_, {1, _}}), do: :discouraged
  def resolve_w(_), do: :neutral

  # -- compile' : Tier x Polarity -> {value, weight} ----------------------
  @tier_rank %{critical: 2, important: 1, refining: 0}
  def tier_rank, do: @tier_rank
  def compile_w(:critical, :promote), do: {:required, 2}
  def compile_w(:critical, :prevent), do: {:forbidden, 2}
  def compile_w(:important, :promote), do: {:encouraged, 1}
  def compile_w(:important, :prevent), do: {:discouraged, 1}
  def compile_w(:refining, :promote), do: {:encouraged, 0}
  def compile_w(:refining, :prevent), do: {:discouraged, 0}

  # -- Precedence ladder (gap 1): specificity -> supersession -> evidence --
  def grade_gt({t1, d1}, {t2, d2}), do: {t1, d1} != {t2, d2} and t1 >= t2 and d1 >= d2
  def rung1({s1, _, _}, {s2, _, _}, act) do
    cond do
      strict_subset?(s1, s2) -> if MapSet.member?(s1, act), do: 1, else: 2
      strict_subset?(s2, s1) -> if MapSet.member?(s2, act), do: 2, else: 1
      true -> 0
    end
  end
  defp strict_subset?(a, b), do: MapSet.subset?(a, b) and not MapSet.equal?(a, b)
  def rung2({_, e1, _}, {_, e2, _}), do: (e1 > e2 && 1) || (e2 > e1 && 2) || 0
  def rung3({_, _, g1}, {_, _, g2}), do: (grade_gt(g1, g2) && 1) || (grade_gt(g2, g1) && 2) || 0
  def precedence(n1, n2, act) do
    Enum.find([rung1(n1, n2, act), rung2(n1, n2), rung3(n1, n2)], 0, &(&1 != 0))
  end

  def may_authorize(h), do: h in [:required, :encouraged, :neutral]

  # audit F-6: two-regime means transfer
  # t_nec: necessary means -> full inheritance; t_aux: auxiliary means
  # -> symmetric attenuation (non-identity, so dual-naturality has content)
  def t_nec(h), do: h
  def t_aux(:required), do: :encouraged
  def t_aux(:encouraged), do: :encouraged
  def t_aux(:neutral), do: :neutral
  def t_aux(:discouraged), do: :discouraged
  def t_aux(:forbidden), do: :discouraged

  # audit F-1: independent reference for resolve_w pinning
  def expected_resolve_w({cs, cw}, {os_, ow}) do
    case resolve({cs, os_}) do
      :neutral when cs == 1 and os_ == 1 ->
        cond do
          cw > ow -> :encouraged
          ow > cw -> :discouraged
          true -> :neutral_gated
        end
      base -> base
    end
  end
end

# ========================= exhaustive verification ========================
defmodule Verify do
  import Deontic
  def law(name, ok, note \\ "") do
    IO.puts("#{if ok, do: "PASS", else: "FAIL"} #{name}#{if note != "", do: " — " <> note, else: ""}")
    ok
  end

  def run do
    hs = h()
    claims = for s <- 0..2, w <- 0..2, do: {s, w}
    scopes = for bits <- 0..7, do: MapSet.new(Enum.filter([:a, :b, :c], fn x ->
               Bitwise.band(bits, Enum.find_index([:a, :b, :c], &(&1 == x)) |> then(&Bitwise.bsl(1, &1))) != 0 end))
    grades = for t <- 0..1, d <- 0..1, do: {t, d}
    norms = for s <- scopes, ep <- 0..2, g <- grades, do: {s, ep, g}
    triples = for n1 <- norms, n2 <- norms, a <- [:a, :b, :c], do: {n1, n2, a}

    results = [
      law("L1 derivation exact", length(axes()) == 5 and Enum.sort(hs) == Enum.sort([:required, :encouraged, :neutral, :discouraged, :forbidden])),
      law("L3 dual involution", Enum.all?(hs, &(dual(dual(&1)) == &1))),
      law("L4 dual antitone", Enum.all?(hs, fn a -> Enum.all?(hs, fn b -> (v(a) <= v(b)) == (v(dual(b)) <= v(dual(a))) end) end)),
      law("L5 fix(dual) = {neutral}", Enum.filter(hs, &(dual(&1) == &1)) == [:neutral]),
      law("L6 Lukasiewicz-5 iso", Enum.all?(hs, &(abs(v(dual(&1)) - (1 - v(&1))) < 1.0e-12))),
      law("L7 fold assoc+comm+idem, resolve.e = id",
        Enum.all?(for x <- claims, y <- claims, z <- claims, do: fold(fold(x, y), z) == fold(x, fold(y, z))) and
        Enum.all?(for x <- claims, y <- claims, do: fold(x, y) == fold(y, x)) and
        Enum.all?(hs, &(resolve(e(&1)) == &1))),
      law("L7d combine order-independent",
        Enum.all?(for a <- hs, b <- hs, c <- hs, do: combine([a, b, c]) == combine([c, b, a])),
        "all 125 triples"),
      law("L7e conflict iff binding-vs-binding",
        Enum.all?(for a <- hs, b <- hs, c <- hs, p = [a, b, c],
          do: (combine(p) == :conflict) == (:required in p and :forbidden in p))),
      law("L9 compile-polarity duality",
        Enum.all?([:critical, :important, :refining], fn t ->
          {hw, w} = compile_w(t, :promote); {ha, ^w} = compile_w(t, :prevent); hw == dual(ha) end),
        "same weight, dual value"),
      law("T7 weighted fold assoc",
        Enum.all?(for x <- claims, y <- claims, z <- claims,
          do: wclaim_fold(wclaim_fold(x, y), z) == wclaim_fold(x, wclaim_fold(y, z)))),
      law("T8 weight breaks (1,1) tie",
        Enum.all?(for w1 <- 0..2, w2 <- 0..2 do
          resolve_w({{1, w1}, {1, w2}}) ==
            cond do w1 > w2 -> :encouraged; w2 > w1 -> :discouraged; true -> :neutral_gated end
        end)),
      law("T9 compile' injective",
        (for t <- [:critical, :important, :refining], p <- [:promote, :prevent], do: compile_w(t, p))
        |> Enum.uniq() |> length() == 6),
      law("T1/T2 precedence total + dual",
        Enum.all?(triples, fn {n1, n2, a} ->
          r = precedence(n1, n2, a)
          r in [0, 1, 2] and r == %{0 => 0, 1 => 2, 2 => 1}[precedence(n2, n1, a)] end),
        "#{length(triples)} conflict triples"),
      law("T4 refusal characterized",
        Enum.all?(triples, fn {n1, n2, a} ->
          {_, e1, g1} = n1; {_, e2, g2} = n2
          (precedence(n1, n2, a) == 0) ==
            (rung1(n1, n2, a) == 0 and e1 == e2 and not grade_gt(g1, g2) and not grade_gt(g2, g1)) end)),
      law("L11 authorization sound",
        not may_authorize(:forbidden) and not may_authorize(:conflict) and
        Enum.all?(hs, &(may_authorize(&1) == (v(&1) >= 0.5)))),
      # -- audit F-2: explicit means transfer, three constrained laws ------
      law("L12 T dual-natural (both regimes)",
        Enum.all?(hs, &(t_nec(dual(&1)) == dual(t_nec(&1)))) and
        Enum.all?(hs, &(t_aux(dual(&1)) == dual(t_aux(&1)))),
        "t_aux != id gives the law force"),
      law("L12b blocking transfer (graded)",
        t_nec(:forbidden) == :forbidden and t_aux(:forbidden) == :discouraged and
        not may_authorize(t_nec(:forbidden)) and not may_authorize(t_aux(:forbidden))),
      law("L12c means never outrank ends, strictly attenuated when auxiliary",
        Enum.all?(hs, &(abs(v(t_nec(&1)) - 0.5) <= abs(v(&1) - 0.5))) and
        Enum.all?(hs, &(abs(v(t_aux(&1)) - 0.5) <= abs(v(&1) - 0.5))) and
        abs(v(t_aux(:required)) - 0.5) < abs(v(:required) - 0.5) and
        abs(v(t_aux(:forbidden)) - 0.5) < abs(v(:forbidden) - 0.5)),
      # -- audit F-1: resolve_w pinned over all 81 states + conflict weight-proof
      law("T11 resolve_w pinned over all states",
        Enum.all?(for c <- claims, o <- claims, do: resolve_w({c, o}) == expected_resolve_w(c, o)),
        "all 81 states incl. every strength-2 branch"),
      law("T12 weight never breaks conflict",
        Enum.all?(for w1 <- 0..2, w2 <- 0..2, do: resolve_w({{2, w1}, {2, w2}}) == :conflict))
    ]

    if Enum.all?(results) do
      IO.puts("\nRESULT: all #{length(results)} laws hold by total enumeration (BEAM embodiment agrees with the Python certificate).")
      System.halt(0)
    else
      IO.puts("\nRESULT: FAILURE — model refuted.")
      System.halt(1)
    end
  end
end

Verify.run()

File: priv/deontic/deontic.py
#!/usr/bin/env python3
"""SPEC_deontic_core — executable formal model, exhaustively verified.

The five deontic values (H) are DERIVED, not postulated, from a three-axis
decomposition; H embeds in an aggregation monoid M whose fold is associative
(the naive 5-valued combination is not — L7c documents the counterexample);
resolution `resolve` maps folded state back to H or to conflict
(-> unresolved_conflict, fail closed).

Finite domain => every law below is checked by total enumeration.
Exit 0 iff all laws hold. Stdlib only. Deterministic.

Vocabulary note: this model is stated in neutral deontic-logic terms because
it is applied to isomorphisms internal to this system. Structural coincidence
with any external normative tradition is a remark, not a claim.
"""
from itertools import product

# ---- Derivation: three axes -> exactly five inhabitants -----------------
# A1 mode: Demand | Option; A2 direction (Demand only): Act | Refrain;
# A3 bindingness (Demand only): Binding | NonBinding.
AXES = [("Demand", d, b) for d in ("Act", "Refrain") for b in (True, False)] + [("Option",)]
NAME = {("Demand", "Act", True): "required", ("Demand", "Act", False): "encouraged",
        ("Demand", "Refrain", True): "forbidden", ("Demand", "Refrain", False): "discouraged",
        ("Option",): "neutral"}
H = tuple(NAME[a] for a in AXES)  # the five deontic values

# ---- Valence chain and involution --------------------------------------
V = {"forbidden": 0.0, "discouraged": 0.25, "neutral": 0.5,
     "encouraged": 0.75, "required": 1.0}
DUAL = {"required": "forbidden", "encouraged": "discouraged", "neutral": "neutral",
        "discouraged": "encouraged", "forbidden": "required"}  # act-negation transfer

# ---- Aggregation monoid M = ({0,1,2}^2, coordinatewise max) ------------
# state (c, o): strongest claim toward commission / omission seen so far.
E = {"required": (2, 0), "encouraged": (1, 0), "neutral": (0, 0),
     "discouraged": (0, 1), "forbidden": (0, 2)}  # embedding e : H -> M
M = list(product(range(3), repeat=2))

def fold(x, y):  # associative by construction
    return (max(x[0], y[0]), max(x[1], y[1]))

def resolve(s):  # resolution M -> H | CONFLICT
    c, o = s
    if c == 2 and o == 2: return "CONFLICT"   # opposed binding demands
    if c == 2: return "required"              # binding beats non-binding
    if o == 2: return "forbidden"
    if c == 1 and o == 1: return "neutral"    # tie -> neutral, gate-annotated
    if c == 1: return "encouraged"
    if o == 1: return "discouraged"
    return "neutral"

def combine(hs):  # multiset combination on H
    s = (0, 0)
    for h in hs: s = fold(s, E[h])
    return resolve(s)

# ---- compile : Tier x Polarity -> H (from the objectives overlay) ------
COMPILE = {("critical", "promote"): "required", ("critical", "prevent"): "forbidden",
           ("important", "promote"): "encouraged", ("important", "prevent"): "discouraged",
           ("refining", "promote"): "encouraged", ("refining", "prevent"): "discouraged"}
TIER_RANK = {"critical": 2, "important": 1, "refining": 0}

# ---- Authorization semantics -------------------------------------------
def may_authorize(h):  # A6 / prohibition-default handled upstream
    return h in {"required", "encouraged", "neutral"}

# ======================= exhaustive verification =========================
checks = []
def law(name, ok, note=""):
    checks.append((name, ok, note))
    print(f"{'PASS' if ok else 'FAIL'} {name}" + (f" — {note}" if note else ""))

law("L1 derivation exact", len(AXES) == 5 and set(H) == set(V) and len(set(H)) == 5,
    "three axes yield exactly five deontic values, bijectively")
law("L2 valence chain total order", sorted(V.values()) == [0.0, .25, .5, .75, 1.0])
law("L3 dual involution", all(DUAL[DUAL[h]] == h for h in H))
law("L4 dual antitone", all((V[a] <= V[b]) == (V[DUAL[b]] <= V[DUAL[a]]) for a in H for b in H))
law("L5 fix(dual) = {neutral}", [h for h in H if DUAL[h] == h] == ["neutral"])
law("L6 Lukasiewicz-5 iso", all(abs(V[DUAL[h]] - (1 - V[h])) < 1e-12 for h in H),
    "v(dual x) = 1 - v(x): (H, chain, dual) ~ L5 De Morgan chain")
law("L7a fold assoc+comm+idem on M", all(fold(fold(a,b),c) == fold(a,fold(b,c))
    for a in M for b in M for c in M) and all(fold(a,b) == fold(b,a) for a in M for b in M)
    and all(fold(a,a) == a for a in M))
law("L7b resolve . e = id", all(resolve(E[h]) == h for h in H),
    "embedding is resolution-faithful")
law("L7c naive 5-valued combine non-assoc witness",
    combine(["encouraged", "encouraged", "discouraged"]) == "neutral" and
    resolve(fold(E["encouraged"], E[resolve(fold(E["encouraged"], E["discouraged"]))])) == "encouraged",
    "hence aggregate in M, resolve once — order-independence restored")
law("L7d combine order-independent", all(
    combine(p) == combine(list(reversed(p)))
    for p in product(H, repeat=3)), "checked over all 125 triples")
law("L7e conflict iff binding-vs-binding", all(
    (combine(list(p)) == "CONFLICT") == ({"required", "forbidden"} <= set(p))
    for p in product(H, repeat=3)), "conflict exactly when binding demands oppose")
law("L8 identity neutral", all(combine([h, "neutral"]) == h for h in H))
law("L9 compile-polarity duality", all(COMPILE[(t, "promote")] == DUAL[COMPILE[(t, "prevent")]]
    for t in TIER_RANK), "polarity commutes with dual: promote/prevent land on dual values")
law("L10 compile monotone strength", all(
    abs(V[COMPILE[(t, d)]] - .5) <= abs(V[COMPILE[(u, d)]] - .5)
    for d in ("promote", "prevent") for t in TIER_RANK for u in TIER_RANK
    if TIER_RANK[t] <= TIER_RANK[u]), "higher tier => stronger (farther from neutral)")
law("L11 authorization sound",
    not may_authorize("forbidden") and not may_authorize("CONFLICT")
    and all(may_authorize(h) == (V[h] >= 0.5) for h in H),
    "forbidden and unresolved conflict never authorize; discouraged gates, not blocks")

# ---- Means transfer (audit F-6): two regimes ---------------------------
# T_NEC: necessary/determinate means -> full inheritance (identity).
#        Blocking transfer at full strength.
# T_AUX: auxiliary/non-determinate means -> symmetric attenuation: one strict
#        step toward neutral on the binding poles. Non-identity, so
#        dual-naturality below has content: a corrupted DUAL now fails L12.
T_NEC = {h: h for h in H}
T_AUX = {"required": "encouraged", "encouraged": "encouraged", "neutral": "neutral",
         "discouraged": "discouraged", "forbidden": "discouraged"}

law("L12 T dual-natural (both regimes)", all(
    Tm[DUAL[h]] == DUAL[Tm[h]] for Tm in (T_NEC, T_AUX) for h in H),
    "transfer commutes with act-negation; T_AUX != id gives the law force")
law("L12b blocking transfer (graded)",
    T_NEC["forbidden"] == "forbidden" and T_AUX["forbidden"] == "discouraged"
    and not may_authorize(T_NEC["forbidden"]) and not may_authorize(T_AUX["forbidden"]),
    "determinate means to a forbidden end is forbidden; auxiliary means is discouraged — neither authorizes")
law("L12c means never outrank ends, strictly attenuated when auxiliary", all(
    abs(V[Tm[h]] - .5) <= abs(V[h] - .5) for Tm in (T_NEC, T_AUX) for h in H)
    and abs(V[T_AUX["required"]] - .5) < abs(V["required"] - .5)
    and abs(V[T_AUX["forbidden"]] - .5) < abs(V["forbidden"] - .5),
    "no-strengthening for both; T_AUX strictly weakens the binding poles")

if all(ok for _, ok, _ in checks):
    print(f"\nRESULT: all {len(checks)} laws hold by total enumeration over |H|=5, |M|=9.")
    raise SystemExit(0)
print("\nRESULT: FAILURE — model refuted; spec must not be graded CANDIDATE.")
raise SystemExit(1)
